File: app-android-biometric-passport-zk/app/src/main/java/org/iranUnchained/utils/VoteSMTInputsBuilder.kt
package org.iranUnchained.utils

import java.math.BigInteger

/**
 * Pure function to build the circuit inputs JSON for voteSMT proof generation.
 * Extracted from VoteSubmissionService for testability.
 */
object VoteSMTInputsBuilder {

    fun buildJson(
        registrationRootHex: String,
        currentDate: BigInteger,
        proposalEventId: BigInteger,
        nullifier: BigInteger,
        secretKey: BigInteger,
        citizenship: BigInteger,
        identityCreationTimestamp: BigInteger,
        votes: List<BigInteger>,
        proposalId: BigInteger
    ): String {
        val votesStr = votes.joinToString(",") { "\"$it\"" }

        return """
        {
            "root": "$registrationRootHex",
            "currentDate": "${currentDate}",
            "proposalEventId": "${proposalEventId}",
            "nullifier": "${nullifier}",
            "secretKey": "${secretKey}",
            "citizenship": "${citizenship}",
            "identityCreationTimestamp": "${identityCreationTimestamp}",
            "vote": [${votesStr}],
            "proposalId": "${proposalId}"
        }
        """.trimIndent()
    }
}

File: app-ios-biometric-passport-zk/IranUnchained/Code/Utils/Web3/VoteSMTInputsBuilder.swift
//
//  VoteSMTInputsBuilder.swift
//  IranUnchained
//
//  Builds circuit inputs JSON for voteSMT (Semaphore-style SMT voting) proof generation.
//  Direct port of Android's VoteSMTInputsBuilder.kt.
//

import Foundation

enum VoteSMTInputsBuilder {

    /// Build the circuit inputs JSON string.
    ///
    /// The vote_smt circuit proves knowledge of a secret + Merkle path in the registration SMT.
    /// Identity fields (citizenship, timestamps) are committed into the SMT leaf during registration.
    static func buildJSON(
        root: String,
        nullifierHash: String,
        nullifier: String,
        vote: String,
        secret: String,
        pathElements: [String],
        pathIndices: [String]
    ) -> String {
        let pathElementsStr = pathElements.map { "\"\($0)\"" }.joined(separator: ",")
        let pathIndicesStr = pathIndices.map { "\"\($0)\"" }.joined(separator: ",")

        return """
        {
            "root": "\(root)",
            "nullifierHash": "\(nullifierHash)",
            "nullifier": "\(nullifier)",
            "vote": "\(vote)",
            "secret": "\(secret)",
            "pathElements": [\(pathElementsStr)],
            "pathIndices": [\(pathIndicesStr)]
        }
        """
    }
}
